### ISSUE TITLE
Update ANGLE_instanced_arrays extension and conformance tests for out-of-range index behavior

### ISSUE THREAD

**Issue by kenrussell** [KhronosGroup/WebGL#835]
Per resolution in the WebGL working group, the strict validation of indices in drawElements calls has been relaxed to allow the behavior specified by the KHR_robust_buffer_access_behavior extension to OpenGL ES 2.0.

The ANGLE_instanced_arrays extension needs to be updated to take this change into consideration.

The WebGL conformance tests also need to be updated to take this change into consideration. Specifically, it is essential that the relaxed behavior be tested as thoroughly as possible.

**Comment by null77** [comment]
Ken, I was looking at removing some checks in ANGLE for range validation. I noticed when I did this several WebGL tests were failing:

WebglConformance_conformance_more_functions_drawArrays
WebglConformance_conformance_more_functions_drawArraysOutOfBounds
WebglConformance_conformance_rendering_draw_arrays_out_of_bounds
WebglConformance_conformance_rendering_out_of_bounds_index_buffers

I looked at some of the tests and at least one was expecting GL_INVALID_OPERATION when drawing with an empty buffer. It looks like the spec doesn't specify this as a guaranteed error. Rather it's okay to have robust access behaviours. Do these tests still need to be updated before ANGLE can remove its array buffer range checks? When robust access is available that is.

**Comment by jdashg** [comment]
Make sure that the state isn't otherwise invalid. I thought I fixed most of
the tests that had valid state except for bad buffer sizes during some work
I did on Firefox's validation.

**Comment by null77** [comment]
I think the drawElements tests might be mostly or entirely resolved. The drawArrays tests are not. I'm not sure if I should open a separate issue for that. But I'll follow up. See [1] for a clear example:

PASS getError was expected value: INVALID_OPERATION : after evaluating: gl.drawArrays(gl.TRIANGLES, 0, 10000)

This out of bounds draw should be able to succeed and produce values from anywhere in the enabled buffers or zero values.

[1]: https://www.khronos.org/registry/webgl/sdk/tests/js/tests/out-of-bounds-test.js

**Comment by kenrussell** [comment]
Looking back at the spec and this conformance test, the reason these drawArrays tests still require INVALID_OPERATION is that it should be inexpensive to ensure that the vertex fetches all fit within the bound buffers; they're just a bunch of quick inequality checks. The same isn't true for drawElements, because of the indirection through the indices.

Is this validation a hot spot? Unless it is expensive, it is better to provide clear errors to the developer wherever possible, so I would prefer not to relax these tests. (And perhaps we should tighten the language in https://www.khronos.org/registry/webgl/specs/latest/1.0/#6.6 to disallow the robust buffer access behavior for drawArrays calls?)

**Comment by null77** [comment]
Yes, this is a hotspot. A bunch of quick inequality checks is not quick. :)

ANGLE caches as much as it can so draw calls are optimized. This reduces the draw call check considerably. But that has a big cost in vertex array state changes. And vertex array changes are a common use pattern between draws. The hotspot is compounded by the checked math necessary to guard against overflow. All the parameters are unsafe. See [this code in angle](https://chromium.googlesource.com/angle/angle/+/master/src/libANGLE/VertexAttribute.cpp#105) for an example. Note it includes heavyweight operations like divide. I don't think there any alternative to this method if you want optimal draw call validation.

The WebGL spec currently has wording ([link](https://www.khronos.org/registry/webgl/specs/latest/1.0/#ATTRIBS_AND_RANGE_CHECKING))based on ARB_robust_buffer_access_behavior. This includes drawElements and drawArrays. Changing DrawArrays for WebGL would be inconsistent with the extension and give worse performance. I recommend we  keep the current spec and update the tests.

**Comment by jdashg** [comment]
Firefox caches these for given vao configurations, so I don't see them as
hotspots for us.

**Comment by kenrussell** [comment]
@null77 your point is taken regarding differing behavior between drawArrays and drawElements. Can you propose a change to the WebGL conformance suite relaxing the drawArrays test, as well as ensuring that the robust buffer access behavior for out-of-range drawArrays calls is tested? Could you also provide performance numbers that motivate this change?

@jdashg any objections to moving in this direction, assuming the performance numbers look good?

**Comment by null77** [comment]
Sounds good @kenrussell . I just ran some tests with ANGLE's GL and Vulkan back-ends. I've found that with one attribute, for a test that just changes the vertex arrays and draws, the amount of time spent in ANGLE is reduced by 13-15% without the validation. This scales to a 26-27% improvement as the number of attributes increases to five. It's likely for eight attributes it would be in the 30%+ range.

Detailed results here: https://docs.google.com/spreadsheets/d/1Bwb1rSSkOcvg1KhwwkcDRkSRHuHWtyfF-62kXivWMDg

I'll move forward with changing the tests with the expectation that this is acceptable.

**Comment by jdashg** [comment]
Sure, I'm fine with it. I suspect the behavior of best-practice code with VAOs will be minimally improved, however.

**Comment by null77** [comment]
PR landed here: https://github.com/KhronosGroup/WebGL/pull/2771

I think you can resolve this issue as fixed. The index out of range tests are already entirely updated.
